airflow/contrib/hooks/gcp_cloudml_hook.py

```diff
@@ -0,0 +1,166 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+import logging
+import random
+import time
+from airflow import settings
+from airflow.contrib.hooks.gcp_api_base_hook import GoogleCloudBaseHook
+from apiclient.discovery import build
+from apiclient import errors
+from oauth2client.client import GoogleCredentials
+
+logging.getLogger('GoogleCloudML').setLevel(settings.LOGGING_LEVEL)
+
+
+def _retry_with_exponential_delay(request, max_n, is_done_func, is_error_func):
```

Comment:
One thing to consider: Airflow operators, themselves, have a retry setting that developers can use in the case of a failure. If you simply allow failures to propagate up, the task will fail, and the developer can decide whether the task should retry. It might be cleaner/easier to remove this retry logic, and just allow failures to go up the stack to Airflow, where config can dictate how to handle failures.

It is nice to have slightly more sophisticated logic, as you do, to handle retry HTTP error codes, and fail outright on others, so I could see an argument for keeping this. Just something to consider.